hand-eye-calibration/full_hand_eye_sample.py

```diff
@@ -0,0 +1,486 @@
+                    )
+
+                    # Using patlib to save images
+                    _save_zdf_and_pose(save_dir, image_num, frame, transform)
+                    print("Image and pose saved")
+
+                time.sleep(0.1)
+
+    _write_robot_state(con, input_data, finish_capture=False, camera_ready=False)
+    time.sleep(1.0)
+    con.send_pause()
+    con.disconnect()
+
+    print(f"Data saved to: {save_dir}")
+
+    return save_dir
+
+
+def perform_hand_eye_calibration(mode: str, data_dir: Path):
+    """ Perform had-eye calibration based on mode
+
+    Args:
+        mode: Calibration mode, eye-in-hand or eye-to-hand
+        data_dir: Path to dataset
+
+    Returns:
+        transform: 4x4 transformation matrix
+        residual: List of residuals
+
+    Raises:
+        RuntimeError: If no feature points are detected
+        ValueError: If calibration mode is invalid
+    """
+
+    calibration_inputs = []
+    idata = 1
+    while True:
+        frame_file = data_dir / f"img{idata:02d}.zdf"
+        pose_file = data_dir / f"pos{idata:02d}.yaml"
+
+        if frame_file.is_file() and pose_file.is_file():
+
+            print(f"Detect feature points from img{idata:02d}.zdf")
+            point_cloud = zivid.Frame(frame_file).get_point_cloud()
+            detected_features = zivid.handeye.detect_feature_points(point_cloud)
+
+            if not detected_features:
+                raise RuntimeError(
+                    f"Failed to detect feature points from frame {frame_file}"
+                )
+
+            print(f"Read robot pose from pos{idata:02d}.yaml")
+            with open(pose_file) as file:
+                pose = pose_from_datastring(file.read())
+
+            detection_result = zivid.handeye.CalibrationInput(pose, detected_features)
+            calibration_inputs.append(detection_result)
+        else:
+            break
+
+        idata += 1
+
+    print(f"\nPerform {mode} calibration")
+    if mode == "eye-in-hand":
+        calibration_result = zivid.handeye.calibrate_eye_in_hand(calibration_inputs)
+    elif mode == "eye-to-hand":
+        calibration_result = zivid.handeye.calibrate_eye_to_hand(calibration_inputs)
+    else:
+        raise ValueError(f"Invalid calibration mode: {mode}")
+
+    transform = calibration_result.hand_eye_transform
+    residuals = calibration_result.per_pose_calibration_residuals
+
+    print("\n\nTransform: \n")
+    np.set_printoptions(precision=5, suppress=True)
+    print(transform)
+
+    print("\n\nResiduals: \n")
+    for num, res in enumerate(residuals, start=1):
+        print(
+            f"pose: {num:02d}   Rotation: {res.rotation:.6f}   Translation: {res.translation:.6f}"
+        )
+
+    return transform, residuals
+
+
+def _main():
+
+    host = "192.168.1.246"
+    port = 30004
+    con, input_data = _initialize_robot_sync(host, port)
+    con.send_start()
+
+    dataset_dir = _generate_dataset(con, input_data)
```

Comment:
Is there a good reason to save the dataset to files just to immediately re-load? Why not just pass it with variables in memory?

Reply:
It would make sense if these were two different scripts. One for generating the dataset, and one for performing the calibration computation.

Reply:
Not its a user input if they want to do generate dataset, do hand-eye or both

Reply:
How would the user choose this?